### PR TITLE
Interpreter: Run interpreter test without exchange executors

### PR DESCRIPTION
**Pull request by ywqzzy**
### What problem does this PR solve?

Issue Number: ref #4609 

Problem Summary:

### What is changed and how it works?
Mock TableScanBlockInputStreams for interpreter tests.
Add stream size configure for user to choose the concurrency of inputstreams.
Provide TiDBTableScan-->schema function, schema->column function.
### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Flash/Coprocessor/DAGContext.h [hunk 1]
```diff
@@ -154,7 +154,6 @@ class DAGContext
         initExecutorIdToJoinIdMap();
         initOutputInfo();
     }
-
     // for test
     explicit DAGContext(UInt64 max_error_count_)
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGContext.h:158]: also `is_test(true)`?

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp [hunk 2]
```diff
@@ -114,7 +116,7 @@ AnalysisResult analyzeExpressions(
     AnalysisResult res;
     ExpressionActionsChain chain;
     // selection on table scan had been executed in handleTableScan
-    if (query_block.selection && !query_block.isTableScanSource())
+    if (query_block.selection && (!query_block.isTableScanSource() || context.getDAGContext()->isTest()))
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:119]: add some annotations like
````
In test mode, filter is not pushed down to table scan
````

File: dbms/src/Flash/Coprocessor/InterpreterDAG.cpp [hunk 3]
```diff
@@ -30,6 +30,8 @@ InterpreterDAG::InterpreterDAG(Context & context_, const DAGQuerySource & dag_)
         max_streams = settings.max_threads;
     else
         max_streams = 1;
+    if (dagContext().isTest())
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/InterpreterDAG.cpp:33]: how about
```suggestion
    else if (dagContext().isTest())
```

File: dbms/src/Storages/Transaction/TiDB.cpp [hunk 4]
```diff
@@ -1103,4 +1104,62 @@ ColumnInfo fieldTypeToColumnInfo(const tipb::FieldType & field_type)
...
+NamesAndTypes genNamesAndTypesFromTableScan(const tipb::TableScan & table_scan)
+{
+    NamesAndTypes names_and_types;
+    names_and_types.reserve(table_scan.columns_size());
+
+    for (Int32 i = 0; i < table_scan.columns_size(); ++i)
+    {
+        String name = "mock_table_scan_" + std::to_string(i);
+        TiDB::ColumnInfo column_info;
+        auto const & ci = table_scan.columns(i);
```
Comment (SeaRise) [inline dbms/src/Storages/Transaction/TiDB.cpp:1133]: use `TiDBTableScan.getColumns()` directly?

File: dbms/src/Storages/Transaction/TiDB.h [hunk 5]
```diff
@@ -398,4 +401,14 @@ String genJsonNull();
...
 
+using ColumnsWithTypeAndName = DB::ColumnsWithTypeAndName;
+using NamesAndTypes = DB::NamesAndTypes;
+using DAGColumnInfo = std::pair<String, ColumnInfo>;
+using DAGSchema = std::vector<DAGColumnInfo>;
+
+DAGSchema genSchemaFromTableScan(const tipb::TableScan & table_scan);
+NamesAndTypes genNamesAndTypesFromTableScan(const tipb::TableScan & table_scan);
+ColumnsWithTypeAndName getColumnWithTypeAndName(const DAGSchema & schema);
+ColumnsWithTypeAndName getColumnWithTypeAndName(const NamesAndTypes & names_and_types);
```
Comment (SeaRise) [inline dbms/src/Storages/Transaction/TiDB.h:412]: We'd better put these methods in separate files.
In my opinion, only
`TiDBColumnInfo toTiDBColumnInfo(tipb::ColumnInfo)` fits here.

File: dbms/src/TestUtils/tests/gtest_mock_executors.cpp [hunk 6]
```diff
@@ -143,6 +142,7 @@ try
         String expected = "topn_1 | order_by: {(<0, String>, desc: false)}, limit: 10\n"
                           " table_scan_0 | {<0, String>, <1, String>}\n";
         ASSERT_DAGREQUEST_EQAUL(expected, request);
+        // executeInterpreter(request, context.context);
```
Comment (SeaRise) [inline dbms/src/TestUtils/tests/gtest_mock_executors.cpp:145]: useless code?

File: dbms/src/Storages/Transaction/TiDB.cpp (hunk 4 above)
Comment (ywqzzy) [reply dbms/src/Storages/Transaction/TiDB.cpp:1133]: ok, maybe I will add a new overload of the function.

File: dbms/src/Storages/Transaction/TiDB.h (hunk 5 above)
Comment (ywqzzy) [reply dbms/src/Storages/Transaction/TiDB.h:412]: Maybe add them in ColumnWithTypeAndName.h and NamesAndTypes.h will make sense?

File: dbms/src/Storages/Transaction/TiDB.cpp [hunk 7]
```diff
@@ -25,6 +25,7 @@
 #include <Storages/Transaction/Collator.h>
 #include <Storages/Transaction/SchemaNameMapper.h>
 #include <Storages/Transaction/TiDB.h>
+#include <Storages/Transaction/TypeMapping.h>
```
Comment (SeaRise) [inline dbms/src/Storages/Transaction/TiDB.cpp]: seems useless

File: dbms/src/Flash/Coprocessor/GenSchemaAndColumn.cpp [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        column_info.name = name;
+        schema.push_back({name, column_info});
+    }
+    return schema;
+}
+
+NamesAndTypes genNamesAndTypesFromTableScan(const tipb::TableScan & table_scan)
+{
+    NamesAndTypes names_and_types;
+    names_and_types.reserve(table_scan.columns_size());
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/GenSchemaAndColumn.cpp]: Can you use `TiDBTableScan.getColumnSize` directly?

File: dbms/src/Flash/Coprocessor/GenSchemaAndColumn.cpp [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+    for (Int32 i = 0; i < table_scan.columns_size(); ++i)
+    {
+        String name = "mock_table_scan_" + std::to_string(i);
+        TiDB::ColumnInfo column_info;
+        auto const & ci = table_scan.columns(i);
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/GenSchemaAndColumn.cpp]: Can you use `TiDBTableScan.getColumns[i]` directly?

File: dbms/src/Flash/Coprocessor/TiDBTableScan.h [hunk 10]
```diff
@@ -50,6 +48,10 @@ class TiDBTableScan
     {
         return table_scan->executor_id();
     }
+    const tipb::Executor * getTableScan() const
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/TiDBTableScan.h:51]: It seems that this method can be avoided

File: dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp [hunk 11]
```diff
@@ -1042,7 +1058,10 @@ void DAGQueryBlockInterpreter::executeImpl(DAGPipeline & pipeline)
     else if (query_block.isTableScanSource())
     {
         TiDBTableScan table_scan(query_block.source, dagContext());
-        handleTableScan(table_scan, pipeline);
+        if (context.getDAGContext()->isTest())
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp:1061]:
```suggestion
        if (dagContext().isTest())
```

File: dbms/src/Flash/Coprocessor/GenSchemaAndColumn.cpp [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+NamesAndTypes genNamesAndTypesFromTableScan(const tipb::TableScan & table_scan)
+{
+    NamesAndTypes names_and_types;
+    names_and_types.reserve(table_scan.columns_size());
+
+    for (Int32 i = 0; i < table_scan.columns_size(); ++i)
+    {
+        String name = "mock_table_scan_" + std::to_string(i);
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/GenSchemaAndColumn.cpp:79]: use fmt?

File: dbms/src/Flash/Coprocessor/TiDBTableScan.h (hunk 10 above)
Comment (ywqzzy) [reply dbms/src/Flash/Coprocessor/TiDBTableScan.h:51]: Only have to implement genSchema and genNamesAndTypes functions with the TiDBTableScan struct as the input? I will fix it.

Comment (SeaRise) [reply dbms/src/Flash/Coprocessor/TiDBTableScan.h]: yes

File: dbms/src/Flash/Coprocessor/GenSchemaAndColumn.cpp [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+        auto const & ci = table_scan.getColumns()[i];
+        column_info.tp = static_cast<TiDB::TP>(ci.tp());
+        column_info.id = ci.column_id();
+        auto type = getDataTypeByColumnInfoForComputingLayer(column_info);
+        names_and_types.push_back({name, type});
+    }
+    return names_and_types;
+}
+
+ColumnsWithTypeAndName getColumnWithTypeAndName(const DAGSchema & schema)
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/GenSchemaAndColumn.cpp:54]: seems useless

File: dbms/src/Flash/Coprocessor/GenSchemaAndColumn.cpp [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include <Flash/Coprocessor/GenSchemaAndColumn.h>
+#include <Storages/Transaction/TypeMapping.h>
+#include <fmt/core.h>
+namespace DB
+{
+DAGSchema genSchema(const TiDBTableScan & table_scan)
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/GenSchemaAndColumn.cpp:19]: seems useless

File: dbms/src/Flash/Coprocessor/GenSchemaAndColumn.cpp [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+
+    for (Int32 i = 0; i < table_scan.getColumnSize(); ++i)
+    {
+        String name = fmt::format("mock_table_scan_{}", i);
+        TiDB::ColumnInfo column_info;
+        auto const & ci = table_scan.getColumns()[i];
+        column_info.tp = static_cast<TiDB::TP>(ci.tp());
+        column_info.id = ci.column_id();
+        auto type = getDataTypeByColumnInfoForComputingLayer(column_info);
+        names_and_types.push_back({name, type});
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/GenSchemaAndColumn.cpp:49]: `TiDBPkColumnID`, `ExtraTableIDColumnID`, `ExtraTableIDColumnID` need to be considered.

like
https://github.com/pingcap/tiflash/blob/dec116f1e9c0caa870b704fad650a5dc1469d6ef/dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp#L603-L655

File: dbms/src/Flash/Coprocessor/GenSchemaAndColumn.cpp [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+        }
+        else
+        {
+            auto type = getDataTypeByColumnInfoForComputingLayer(column_info);
+            NameAndTypePair pair = {name, type};
+            names_and_types.push_back(std::move(pair));
+        }
+    }
+    return names_and_types;
+}
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/GenSchemaAndColumn.cpp:51]:
```suggestion
NamesAndTypes genNamesAndTypes(const TiDBTableScan & table_scan)
{
    NamesAndTypes names_and_types;
    names_and_types.reserve(table_scan.getColumnSize());

    for (Int32 i = 0; i < table_scan.getColumnSize(); ++i)
    {
        TiDB::ColumnInfo column_info;
        const auto & ci = table_scan.getColumns()[i];
        column_info.tp = static_cast<TiDB::TP>(ci.tp());
        column_info.cid = ci.column_id();
    
        switch (column_info.cid)
        {
        case TiDBPkColumnID:
            names_and_types.emplace_back(MutableSupport::tidb_pk_column_name, getDataTypeByColumnInfoForComputingLayer(column_info));
            break;
        case ExtraTableIDColumnID:
            names_and_types.emplace_back(MutableSupport::extra_table_id_column_name, MutableSupport::extra_table_id_column_type);
            break;
        default:
            names_and_types.emplace_back(fmt::format("mock_table_scan_{}", i), getDataTypeByColumnInfoForComputingLayer(column_info));
        }
    }
    return names_and_types;
}
```

File: dbms/src/Flash/Coprocessor/GenSchemaAndColumn.cpp [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+    {
+        TiDB::ColumnInfo column_info;
+        const auto & ci = table_scan.getColumns()[i];
+        column_info.tp = static_cast<TiDB::TP>(ci.tp());
+        column_info.id = ci.column_id();
+
+        switch (column_info.id)
+        {
+        case TiDBPkColumnID:
+            names_and_types.emplace_back(MutableSupport::tidb_pk_column_name, getDataTypeByColumnInfoForComputingLayer(column_info));
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/GenSchemaAndColumn.cpp:33]: add todo/comment about the data type of pk_handle_columns

**Review by SeaRise — APPROVED** [review]

File: dbms/src/Flash/Coprocessor/DAGContext.h [hunk 18]
```diff
@@ -167,8 +167,30 @@ class DAGContext
         , max_recorded_error_count(max_error_count_)
         , warnings(max_recorded_error_count)
         , warning_count(0)
+        , is_test(true)
```
Comment (Lloyd-Pottiger) [inline dbms/src/Flash/Coprocessor/DAGContext.h:170]: How about using macro like `NDEBUG` or `DBMS_PUBLIC_GTEST`?

File: dbms/src/Flash/Coprocessor/GenSchemaAndColumn.cpp [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+    {
+        TiDB::ColumnInfo column_info;
+        const auto & ci = table_scan.getColumns()[i];
+        column_info.tp = static_cast<TiDB::TP>(ci.tp());
+        column_info.id = ci.column_id();
+
+        switch (column_info.id)
+        {
+        case TiDBPkColumnID:
+            // todo: need to check if the type of pk_handle_columns matches the type that used in delta merge tree.
```
Comment (Lloyd-Pottiger) [inline dbms/src/Flash/Coprocessor/GenSchemaAndColumn.cpp:33]: ~~`todo: `~~  
`TODO: `?

Comment (ywqzzy) [inline dbms/src/Flash/Coprocessor/GenSchemaAndColumn.cpp:33]:
```suggestion
            // TODO: need to check if the type of pk_handle_columns matches the type that used in delta merge tree.
```

**Review by Lloyd-Pottiger — APPROVED** [review]

**Review by windtalker — APPROVED** [review]
LGTM

**Pull request merged**